relay-server/src/utils/profile.rs

```diff
@@ -57,28 +76,140 @@ impl AndroidProfile {
     }
 }
 
-#[derive(Debug, Fail)]
-pub enum ProfileError {
-    #[fail(display = "invalid json in profile")]
-    InvalidJson(#[cause] serde_json::Error),
-    #[fail(display = "invalid base64 value")]
-    InvalidBase64Value,
-    #[fail(display = "invalid sampled profile")]
-    InvalidSampledProfile,
-    #[fail(display = "cannot serialize payload")]
-    CannotSerializePayload,
-}
-
 pub fn parse_android_profile(item: &mut Item) -> Result<(), ProfileError> {
     let mut profile: AndroidProfile =
         serde_json::from_slice(&item.payload()).map_err(ProfileError::InvalidJson)?;
 
     profile.parse()?;
 
+    if profile.profile.as_ref().unwrap().events.len() < 2 {
+        return Err(ProfileError::NotEnoughSamples);
+    }
+
+    match serde_json::to_vec(&profile) {
+        Ok(payload) => item.set_payload(ContentType::Json, &payload[..]),
+        Err(_) => return Err(ProfileError::CannotSerializePayload),
+    };
+
+    Ok(())
+}
+
+fn strip_pointer_authentication_code<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: de::Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+    let address_without_prefix = s.trim_start_matches("0x");
+    match u64::from_str_radix(address_without_prefix, 16) {
+        // https://github.com/microsoft/plcrashreporter/blob/748087386cfc517936315c107f722b146b0ad1ab/Source/PLCrashAsyncThread_arm.c#L84
+        Ok(address) => Ok(format!("{:#x}", address & 0x0000000FFFFFFFFF)),
+        Err(err) => Err(de::Error::custom(format!(
+            "Failed to strip pointer authentication code: {}",
+            err
+        ))),
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Frame {
+    function: String,
+
+    #[serde(deserialize_with = "strip_pointer_authentication_code")]
+    instruction_addr: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Sample {
+    frames: Vec<Frame>,
+    queue_address: Option<String>,
+
+    #[serde(deserialize_with = "deserialize_number_from_string")]
+    relative_timestamp_ns: u64,
+
+    thread_id: u64,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ThreadMetadata {
+    name: String,
+    priority: u32,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct QueueMetadata {
+    label: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct SampledProfile {
+    samples: Vec<Sample>,
+    thread_metadata: HashMap<String, ThreadMetadata>,
+    queue_metadata: HashMap<String, QueueMetadata>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Image {
```

Comment:
This and other types are very similar to the existing types in the event protocol. Depending on how much you still want to refactor this, it would be good if we start looking into sharing the definitions.

The main challenge with this is that we base most of our protocol on the [`Annotated`](https://getsentry.github.io/relay/relay_general/types/struct.Annotated.html), which requires its own serializing and deserializing implementation. It's notably slower than plain serde, and we've been long looking to re-align it with serde.

Reply:
I'm willing to refactor things to make them more cohesive. I think we should still ship that though because it sounds like a way larger effort that should really be done on its own as opposed to doing it on top of another change.

Reply:
This won't be written back anywhere. I'll use in for symbolication later and then discard it.

Reply:
If this goes through symbolicator we should find a way to make those structs align and fail compilation if they don't. For the time being I would add yet another roundtrip test that ensures this.

Reply:
I made a test for that but there's a few caveats. The iOS SDK is using a function get generate debug meta that was already part of the SDK and this function gets us `apple` debug images. Two problems from there:
- It seems the Symbolicator wants Macho now, and that's why we have to rename the `uuid` field to `debug_id`, but then the image would be more of type `macho`.
- It doesn't provide some fields that seems required when building an `apple` debug image (`cpu_type`, `cpu_subtype` and `arch`).  I added those fields in our image struct for now.

So, we're investigating to see how to generate `macho` images and then I'll make this test a little bit less weird and will also likely remove the need to rename the `uuid` field to `debug_id`.